# Image Manipulations and Image Spaces

In [1]:
import numpy as np
from PIL import Image

In [2]:
# Read an image
image = Image.open("4.1.06.tiff")

# Show the image
image.show()

In [3]:
# Transform the image to numpy array to manipulate it
image_array = np.array(image)

# Inspect the image array
print(f"Shape: {image_array.shape}")
print(f"Red channel:\n{image_array[:, :, 0]}")
print(f"Green channel:\n{image_array[:, :, 1]}")
print(f"Blue channel:\n{image_array[:, :, 2]}")

Shape: (256, 256, 3)
Red channel:
[[233 202 205 ... 233 235 239]
 [230 203 206 ... 232 234 235]
 [230 200 200 ... 232 233 235]
 ...
 [ 48  81  84 ...  75  68  84]
 [ 64 155 115 ...  81  83  85]
 [  0   0   0 ...   0   0   0]]
Green channel:
[[  0 217 218 ... 229 229 229]
 [217 217 217 ... 227 227 229]
 [217 217 216 ... 228 228 229]
 ...
 [ 43  58  43 ...  40  34  40]
 [ 58 155  99 ...  51  45  55]
 [  0   0   0 ...   0   0   0]]
Blue channel:
[[246 221 220 ... 227 227 226]
 [237 221 220 ... 227 225 228]
 [236 221 220 ... 227 227 226]
 ...
 [ 97  70  76 ...  72  72  81]
 [ 94 135  89 ...  84  74  80]
 [  0   0   0 ...   0   0   0]]


In [4]:
# Manipulation example: make the image darker
image_array_v1 = (image_array * 0.6).astype(np.uint8)
image_v1 = Image.fromarray(image_array_v1)
image_v1.show()

In [5]:
# Manipulation example: manipulate the red channel
image_array_v2 = image_array[:, :, 0]
image_v2 = Image.fromarray(image_array_v2)
image_v2.show()
# Why is it grey?
# Because if only one channel is used, the image is considered to be greyscale.

# Manipulation example: set explicitly green and blue channels to 0
image_array_v2 = image_array.copy()
image_array_v2[:, :, 1] = 0  # Set green channel to 0
image_array_v2[:, :, 2] = 0  # Set blue channel to 0
image_v2 = Image.fromarray(image_array_v2)
image_v2.show() # The image has now only the red channel

In [6]:
# Transform RGB array to Lab array
from skimage.color import rgb2lab, lab2rgb

image_array_lab = rgb2lab(image_array)
# Inspect the image array
print(f"Shape: {image_array_lab.shape}")
print(f"L channel:\n{image_array_lab[:, :, 0]}")
print(f"a channel:\n{image_array_lab[:, :, 1]}")
print(f"b channel:\n{image_array_lab[:, :, 2]}")

# Transform Lab array to RGB array
image_array_rgb = lab2rgb(image_array_lab)

# Inspect the image array
print(f"Shape: {image_array_rgb.shape}")
print(f"Red channel:\n{image_array_rgb[:, :, 0]}")
print(f"Green channel:\n{image_array_rgb[:, :, 1]}")
print(f"Blue channel:\n{image_array_rgb[:, :, 2]}")

# Show the image
image = Image.fromarray((image_array_rgb * 255).astype(np.uint8))
image.show()

Shape: (256, 256, 3)
L channel:
[[56.06815251 85.70649804 86.15053474 ... 91.19291207 91.34603556
  91.63142667]
 [88.23904411 85.77686739 85.96353513 ... 90.61551135 90.71930243
  91.37102671]
 [88.21113414 85.56683406 85.28000986 ... 90.86603634 90.94247013
  91.3211776 ]
 ...
 [20.90022822 27.34665366 24.11179235 ... 21.87751504 19.39562593
  23.66076759]
 [26.51742594 63.4755503  43.19728362 ... 26.03168062 24.35542298
  27.4164282 ]
 [ 0.          0.          0.         ...  0.          0.
   0.        ]]
a channel:
[[93.62288486 -4.2523033  -3.9780591  ...  0.99443885  1.67341931
   2.86650061]
 [ 8.01890711 -3.92501161 -3.12997968 ...  1.68553016  2.01711271
   1.8468912 ]
 [ 7.83365831 -4.90474217 -4.5822508  ...  1.17023227  1.50912564
   1.50066387]
 ...
 [18.27126423 12.37713207 24.22905097 ... 21.93389944 23.14911905
  27.02166832]
 [11.83917891 -3.61124745  4.6495671  ... 19.85473825 22.23999884
  17.95949597]
 [ 0.          0.          0.         ...  0.          0.
   0.

In [7]:
# Transform RGB to HSV
from skimage.color import rgb2hsv, hsv2rgb

image_array_hsv = rgb2hsv(image_array)
# Inspect the image array
print(f"Shape: {image_array_hsv.shape}")
print(f"H channel:\n{image_array_hsv[:, :, 0]}")
print(f"S channel:\n{image_array_hsv[:, :, 1]}")
print(f"V channel:\n{image_array_hsv[:, :, 2]}")

# Histogram equalization on the V channel
from skimage.exposure import equalize_hist

image_array_hsv[:, :, 2] = equalize_hist(image_array_hsv[:, :, 2])
print(f"V channel after equalization:\n{image_array_hsv[:, :, 2]}")

# Transform HSV to RGB
image_array_rgb = hsv2rgb(image_array_hsv)

# Show the image
image = Image.fromarray((image_array_rgb * 255).astype(np.uint8))
image.show()

Shape: (256, 256, 3)
H channel:
[[0.82452575 0.53508772 0.52222222 ... 0.05555556 0.04166667 0.03846154]
 [0.775      0.53703704 0.53571429 ... 0.         0.03703704 0.02380952]
 [0.78070175 0.53174603 0.53333333 ... 0.03333333 0.02777778 0.05555556]
 ...
 [0.68209877 0.91304348 0.86585366 ... 0.84761905 0.81578947 0.84469697]
 [0.69444444 0.16666667 0.06410256 ... 0.81818182 0.87280702 0.86111111]
 [0.         0.         0.         ... 0.         0.         0.        ]]
S channel:
[[1.         0.08597285 0.06818182 ... 0.02575107 0.03404255 0.05439331]
 [0.08438819 0.08144796 0.06363636 ... 0.02155172 0.03846154 0.02978723]
 [0.08050847 0.09502262 0.09090909 ... 0.02155172 0.02575107 0.03829787]
 ...
 [0.55670103 0.28395062 0.48809524 ... 0.46666667 0.52777778 0.52380952]
 [0.38297872 0.12903226 0.22608696 ... 0.39285714 0.45783133 0.35294118]
 [0.         0.         0.         ... 0.         0.         0.        ]]
V channel:
[[0.96470588 0.86666667 0.8627451  ... 0.91372549 0.921568

# Compute Power

In [8]:
# Coefficient defines in the slides:
gamma = 0.7755
w_0 = 1.48169521 * 10**(-6)
w_r = 2.13636845 * 10**(-7)
w_g = 1.77746705 * 10**(-7)
w_b = 2.14348309 * 10**(-7)

# First compute the power consumption for each pixel
def compute_power_pixel(r, g, b):
    return w_r * r**gamma + w_g * g**gamma + w_b * b**gamma # From the slides
    
    
def compute_power(image_array_rgb):
    # Receive an image as input and return the power consumption based on the formula on the slides
    
    # Works on the R,G,B space of the image
    # The image is saved in the variable image_array
    
    # Split the image into the three channels    
    red_channel = image_array_rgb[:, :, 0]
    green_channel = image_array_rgb[:, :, 1]
    blue_channel = image_array_rgb[:, :, 2]
    
    
    
    # Apply the function to all pixels
    sum_pixel_power = 0
    height, width = red_channel.shape   # Get the dimensions of the image, assuming all channels have the same dimensions
    for i in range(height):
        for j in range(width):
            power_per_pixel = compute_power_pixel(red_channel[i, j], green_channel[i, j], blue_channel[i, j])
            sum_pixel_power += power_per_pixel
    
    power_image = w_0 + sum_pixel_power
    print('Power consumption of the image: ', power_image)
    return power_image

# compute_power(image_array_rgb)

### Apply trasformation to all the images

#### Darker image only

In [58]:
import os

# Directory containing the images
image_dir = "misc"
image_dir_optimized = "darker"
output_file = "image_power_consumption_dark.txt"

# Open the output file
with open(output_file, "w") as file:
    file.write("Image,Original Power,Optimized Power,Power Saved (%)\n")
    
    # Iterate over all images in the directory
    for image_name in os.listdir(image_dir):
        if image_name.endswith(".tiff"):    # Only process TIFF images
            # Read the image
            image_path = os.path.join(image_dir, image_name)
            image = Image.open(image_path)
            image_array = np.array(image)
            # image.show()
            
            
            # Compute the power consumption of the original image
            power_image_orig = compute_power(image_array)
             
            # Apply the transformation to make the image slightly darker
            image_array_optimized = (image_array * 0.9).astype(np.uint8)
            
            # Show the optimized image
            image_optimized = Image.fromarray(image_array_optimized)
            # image_optimized.show()
            
            # Save the optimized image to disk
            optimized_image_path = os.path.join(image_dir,image_dir_optimized, f"optimized_{image_name}")
            image_optimized.save(optimized_image_path)
            
            # Compute the power consumption of the optimized image
            power_image_optimized = compute_power(image_array_optimized)
            
            # Compute the power saved in percentage
            power_saved_percentage = ((power_image_orig - power_image_optimized) / power_image_orig) * 100
            
            # Write the results to the file
            file.write(f"{image_name},{power_image_orig},{power_image_optimized},{power_saved_percentage}\n")


Power consumption of the image:  0.885606308986581
Power consumption of the image:  0.8093931366191035
Power consumption of the image:  0.5563255685029074
Power consumption of the image:  0.503096583435899
Power consumption of the image:  1.8298231909002254
Power consumption of the image:  1.6813933152966476
Power consumption of the image:  1.5853308824426224
Power consumption of the image:  1.45587979572389
Power consumption of the image:  1.8194199267116073
Power consumption of the image:  1.6718544561884638
Power consumption of the image:  1.7291402637841502
Power consumption of the image:  1.5884718223228302
Power consumption of the image:  2.080625636068071
Power consumption of the image:  1.9125834675679354
Power consumption of the image:  1.9826469295054565
Power consumption of the image:  1.8222774065726717
Power consumption of the image:  5.8809689947281685
Power consumption of the image:  5.3992602864828125
Power consumption of the image:  6.640676209770246
Power consumption 

##### Analysis of the results

In [ ]:
import pandas as pd
from IPython.display import Markdown, display

# Read the file into a pandas dataframe
df_results = pd.read_csv("image_power_consumption_dark.txt")

# Display the dataframe as a markdown table for better visualization
def display_markdown_table(df):
    table = df.to_markdown(index=False)
    display(Markdown(table))

display_markdown_table(df_results)

| Image       |   Original Power |   Optimized Power |   Power Saved (%) |
|:------------|-----------------:|------------------:|------------------:|
| 4.1.01.tiff |         0.885606 |          0.809393 |           8.60576 |
| 4.1.02.tiff |         0.556326 |          0.503097 |           9.56796 |
| 4.1.03.tiff |         1.82982  |          1.68139  |           8.11171 |
| 4.1.04.tiff |         1.58533  |          1.45588  |           8.16556 |
| 4.1.05.tiff |         1.81942  |          1.67185  |           8.11058 |
| 4.1.06.tiff |         1.72914  |          1.58847  |           8.13517 |
| 4.1.07.tiff |         2.08063  |          1.91258  |           8.07652 |
| 4.1.08.tiff |         1.98265  |          1.82228  |           8.08866 |
| 4.2.01.tiff |         5.88097  |          5.39926  |           8.19098 |
| 4.2.03.tiff |         6.64068  |          6.09977  |           8.14529 |
| 4.2.05.tiff |         8.93206  |          8.21361  |           8.04354 |
| 4.2.06.tiff |         6.44016  |          5.91426  |           8.16592 |
| 4.2.07.tiff |         5.85501  |          5.37512  |           8.19626 |
| house.tiff  |         7.80034  |          7.16954  |           8.08682 |

The results show the power consumption of various images before and after optimization, along with the percentage of power saved. Here are some key points from the analysis:

1. **Consistent Power Savings**: The percentage of power saved across all images is relatively consistent, ranging from approximately 8.04% to 9.57%. This indicates that the optimization technique applied is uniformly effective across different images.

2. **Highest and Lowest Savings**:

   - The highest power saving is observed in `4.1.02.tiff` with 9.57%.
   - The lowest power saving is observed in `4.2.05.tiff` with 8.04%.

3. **Power Consumption**:

   - The image with the highest original power consumption is `4.2.05.tiff` at 8.93 units.
   - The image with the lowest original power consumption is `4.1.02.tiff` at 0.56 units.

4. **Optimized Power**:

   - The optimized power consumption is lower for all images, confirming the effectiveness of the optimization process.

5. **Overall Effectiveness**: The optimization process consistently reduces power consumption by around 8-9%, which can be significant in applications where power efficiency is critical.

These results suggest that the optimization technique is robust and can be applied to a variety of images to achieve significant power savings.
Images can be further optimized by acting on the blue channel, as we will see in the next section

For OLED:
- Dynamic quality setting 
- color selection/transformation

We can act only on the color

In [ ]:
from skimage.color import rgb2yuv, yuv2rgb
from PIL import Image
import numpy as np
import os

# Directory containing the images
image_dir = "misc"
image_dir_optimized = "darker_enhanced"
output_file = "image_power_consumption_dark_enhanced.txt"

brightness_factor = 0.5

# Open the output file
with open(output_file, "w") as file:
    file.write("Image,Original Power,Optimized Power,Power Saved (%)\n")
    
    # Iterate over all images in the directory
    for image_name in os.listdir(image_dir):
        if image_name.endswith(".tiff"):    # Only process TIFF images
            # Read the image
            image_path = os.path.join(image_dir, image_name)
            image = Image.open(image_path)
            image_array = np.array(image)
            
            # Compute the power consumption of the original image
            power_image_orig = compute_power(image_array)
             
            # Convert the image from RGB to YUV
            image_array_yuv = rgb2yuv(image_array)

            # Identify pixels with higher power impact (e.g., blue pixels)
            blue_pixels = (image_array[:, :, 2] > image_array[:, :, 0]) & (image_array[:, :, 2] > image_array[:, :, 1])
            non_blue_pixels = ~blue_pixels

            # Reduce the luminance of blue pixels
            image_array_yuv[blue_pixels, 0] *= 0.9

            # # Boost the luminance of non-blue pixels
            # image_array_yuv[non_blue_pixels, 0] = np.clip(image_array_yuv[non_blue_pixels, 0] / brightness_factor, 0, 1)
            

            # Convert the modified YUV image back to RGB
            image_array_optimized = (yuv2rgb(image_array_yuv) * 255).astype(np.uint8)            

            # Show the optimized image
            image_optimized = Image.fromarray(image_array_optimized)
            # image_optimized.show()

            # Save the optimized image to disk
            optimized_image_path = os.path.join(image_dir, image_dir_optimized, f"optimized_{image_name}")
            image_optimized.save(optimized_image_path)

            # Compute the power consumption of the optimized image
            power_image_optimized = compute_power(image_array_optimized)

            # Compute the power saved in percentage
            power_saved_percentage = ((power_image_orig - power_image_optimized) / power_image_orig) * 100
            
            # Write the results to the file
            file.write(f"{image_name},{power_image_orig},{power_image_optimized},{power_saved_percentage}\n")

Power consumption of the image:  0.885606308986581
Power consumption of the image:  0.8765509205793689
Power consumption of the image:  0.5563255685029074
Power consumption of the image:  0.5470541138896168
Power consumption of the image:  1.8298231909002254
Power consumption of the image:  1.7305873407458925
Power consumption of the image:  1.5853308824426224
Power consumption of the image:  1.5131559238006749
Power consumption of the image:  1.8194199267116073
Power consumption of the image:  1.7542865944269452
Power consumption of the image:  1.7291402637841502
Power consumption of the image:  1.6591370841628545
Power consumption of the image:  2.080625636068071
Power consumption of the image:  2.0792247872047604
Power consumption of the image:  1.9826469295054565
Power consumption of the image:  1.9811868573302445
Power consumption of the image:  5.8809689947281685
Power consumption of the image:  6.016723660431499
Power consumption of the image:  6.640676209770246
Power consumptio

##### Analysis of the results

The technique used to enhacned the images is not working very well. The saved power is little in most cases, and in the case of image `4.2.01.tiff` the energy used is even higher than the original image.  

#### Darker image and less blue

In [57]:
import os

# Directory containing the images
image_dir = "misc"
image_dir_optimized = "darker_blue"
output_file = "image_power_consumption_dark_blue.txt"

# Open the output file
with open(output_file, "w") as file:
    file.write("Image,Original Power,Optimized Power,Power Saved (%)\n")
    
    # Iterate over all images in the directory
    for image_name in os.listdir(image_dir):
        if image_name.endswith(".tiff"):    # Only process TIFF images
            # Read the image
            image_path = os.path.join(image_dir, image_name)
            image = Image.open(image_path)
            image_array = np.array(image)
            # image.show()
            
            
            # Compute the power consumption of the original image
            power_image_orig = compute_power(image_array)
             
            # Apply the transformation to make the image slightly darker
            image_array_optimized = (image_array * 0.9).astype(np.uint8)
            # Tweak the blue channel
            image_array_optimized[:, :, 2] = (image_array_optimized[:, :, 2] * 0.9).astype(np.uint8)
            
            # Show the optimized image
            image_optimized = Image.fromarray(image_array_optimized)
            # image_optimized.show()
            # Save the optimized image to disk
            optimized_image_path = os.path.join(image_dir,image_dir_optimized, f"optimized_{image_name}")
            image_optimized.save(optimized_image_path)
            
            # Compute the power consumption of the optimized image
            power_image_optimized = compute_power(image_array_optimized)
            
            # Compute the power saved in percentage
            power_saved_percentage = ((power_image_orig - power_image_optimized) / power_image_orig) * 100
            
            # Write the results to the file
            file.write(f"{image_name},{power_image_orig},{power_image_optimized},{power_saved_percentage}\n")


Power consumption of the image:  0.885606308986581
Power consumption of the image:  0.788168594341055
Power consumption of the image:  0.5563255685029074
Power consumption of the image:  0.48793118528129703
Power consumption of the image:  1.8298231909002254
Power consumption of the image:  1.6319107763588685
Power consumption of the image:  1.5853308824426224
Power consumption of the image:  1.4116455338209875
Power consumption of the image:  1.8194199267116073
Power consumption of the image:  1.6239015779376171
Power consumption of the image:  1.7291402637841502
Power consumption of the image:  1.5400853001873949
Power consumption of the image:  2.080625636068071
Power consumption of the image:  1.8638629748709739
Power consumption of the image:  1.9826469295054565
Power consumption of the image:  1.777314530916956
Power consumption of the image:  5.8809689947281685
Power consumption of the image:  5.273096538282283
Power consumption of the image:  6.640676209770246
Power consumption

In [13]:
# Already imported:
# import pandas as pd
# from IPython.display import Markdown, display


# Read the file into a pandas dataframe
df_results = pd.read_csv("image_power_consumption_dark_blue.txt")

# Display the dataframe as a markdown table for better visualization
def display_markdown_table(df):
    table = df.to_markdown(index=False)
    display(Markdown(table))

display_markdown_table(df_results)

| Image       |   Original Power |   Optimized Power |   Power Saved (%) |
|:------------|-----------------:|------------------:|------------------:|
| 4.1.01.tiff |         0.885606 |          0.788169 |           11.0024 |
| 4.1.02.tiff |         0.556326 |          0.487931 |           12.2939 |
| 4.1.03.tiff |         1.82982  |          1.63191  |           10.8159 |
| 4.1.04.tiff |         1.58533  |          1.41165  |           10.9558 |
| 4.1.05.tiff |         1.81942  |          1.6239   |           10.7462 |
| 4.1.06.tiff |         1.72914  |          1.54009  |           10.9335 |
| 4.1.07.tiff |         2.08063  |          1.86386  |           10.4181 |
| 4.1.08.tiff |         1.98265  |          1.77731  |           10.3565 |
| 4.2.01.tiff |         5.88097  |          5.2731   |           10.3363 |
| 4.2.03.tiff |         6.64068  |          5.93808  |           10.5801 |
| 4.2.05.tiff |         8.93206  |          7.97085  |           10.7614 |
| 4.2.06.tiff |         6.44016  |          5.75349  |           10.6623 |
| 4.2.07.tiff |         5.85501  |          5.26676  |           10.047  |
| house.tiff  |         7.80034  |          6.97707  |           10.5542 |

##### Analysis of the results

The higher power saved in percentage is obtained with the image `4.1.02.tiff`: visually speaking, the change made to the blue channel does not create artifacts, and the image retains more or less the same quality saving 12.2% of energy.

# Compute Distortion

In [19]:
import math

# Convert the image from RGB to Lab space:
image_array_lab = rgb2lab(image_array)

def compute_distortion(original_image, modified_image):
    # To compute the distortion we will work on the L,a,b space of the image
    
    # Split the images into the three channels
    L_channel_original = original_image[:, :, 0]
    a_channel_original = original_image[:, :, 1]
    b_channel_original = original_image[:, :, 2]
    
    L_channel_modified = modified_image[:, :, 0]
    a_channel_modified = modified_image[:, :, 1]
    b_channel_modified = modified_image[:, :, 2]
    
    # Compute the euclidean distance between the two images:
    sum_pixel_distortion = 0
    for i in range(original_image.shape[0]):
        for j in range(original_image.shape[1]):
            # For each pixel compute the euclidean distance
            L_diff = np.subtract(L_channel_original[i, j], L_channel_modified[i, j], dtype=np.float64)
            a_diff = np.subtract(a_channel_original[i, j], a_channel_modified[i, j], dtype=np.float64)
            b_diff = np.subtract(b_channel_original[i, j], b_channel_modified[i, j], dtype=np.float64)
            sum_pixel_distortion += math.sqrt(math.pow(L_diff, 2) + math.pow(a_diff, 2) + math.pow(b_diff, 2))
    
    return sum_pixel_distortion

def compute_percentage_distortion(distortion):
    w = image_array.shape[0]
    h = image_array.shape[1]
    max_distortion = math.sqrt(math.pow(100, 2) + math.pow(255, 2) + math.pow(255, 2)) 
    return (distortion / (w * h * max_distortion)) * 100

# Test the function
print('[TEST] Distortion between the original and itself must be zero, result: ', compute_percentage_distortion(compute_distortion(image_array, image_array)))
print('Distortion between the original and the darker image: ', compute_percentage_distortion(compute_distortion(image_array, image_array_v1)))
print('Distortion between the original and the red image: ', compute_percentage_distortion(compute_distortion(image_array, image_array_v2)))



[TEST] Distortion between the original and itself must be zero, result:  0.0
Distortion between the original and the darker image:  28.350745442646634
Distortion between the original and the red image:  30.18502996870146


In [ ]:
import os

# Directory containing the images
image_dir = "misc"
output_file = "image_distorsion"
optimized_image_dir = ["darker", "darker_blue", "darker_enhanced"]


def open_image(image_path, image_name):
    image = Image.open(os.path.join(image_dir, image_path, f"optimized_{image_name}"))
    image_array = np.array(image)
    return image_array


def files_init():
    for optimized_image_dir_name in optimized_image_dir:
        output_file = f"image_distorsion_{optimized_image_dir_name}.txt"
        with open(output_file, "w") as file:
            file.write("Image,Optimized image,Distorsion,Distorsion (%)\n")


# Iterate over all images in the directory
files_init()    # Initialize the files

for image_name in os.listdir(image_dir):
    if image_name.endswith(".tiff") and not image_name.startswith("optimized"):    # Only process TIFF original images   
        # Read the image
        image_path = os.path.join(image_dir, image_name)
        image = Image.open(image_path)
        image_array = np.array(image)
        # image.show()
        
        for optimized_image_dir_name in optimized_image_dir:
            # Read the optimized image
            optimized_image_array = open_image(optimized_image_dir_name, image_name)
            
            # Compute the distortion between the original and the optimized image
            distortion = compute_distortion(image_array, optimized_image_array)
            distortion_percentage = compute_percentage_distortion(distortion)
            with open(f"image_distorsion_{optimized_image_dir_name}.txt", "a") as file:
                # Write the results to the file
                file.write(f"{image_name},{'optimized_'+image_name},{distortion},{distortion_percentage}\n")
            
            
        



##### Analysis of the results

The results stored in `image_distorsion_dark_blue.txt` provide a detailed analysis of the distortion between the original images and their optimized counterparts.

1. **Distortion Values**:
   - The distortion values vary significantly across different images, with some images experiencing minimal distortion and others experiencing more significant changes.  
   For instance, `4.1.02.tiff` has a distortion percentage of approximately 2.32%, while `4.2.05.tiff` has a distortion percentage of approximately 12.08%. This suggests that the optimization process is more effective in preserving the quality of some images compared to others.

1. **Overall Trends**:
   - The data shows that the optimization process generally introduces some level of distortion, but the extent of this distortion is image-dependent.
   - Images with higher original power consumption tend to have higher distortion values, indicating a trade-off between power optimization and image quality.

2. **Specific Observations**:
   - `4.1.02.tiff` has one of the lowest distortion percentages (2.32%), suggesting that the optimization process is highly effective for this image with minimal quality loss.
   - `4.2.05.tiff` has the highest distortion percentage (12.08%), indicating that the optimization process significantly alters this image.

In conclusion, while the optimization process effectively reduces power consumption, it introduces varying levels of distortion across different images. The results suggest a need for a balanced approach to optimize power consumption while minimizing image quality loss. Further analysis could focus on understanding the factors contributing to higher distortion in certain images and refining the optimization process to achieve better overall results.

# DVS

In [ ]:
def compute_pixel_current():
    raise NotImplementedError

In [ ]:
def compute_panel_power():
    raise NotImplementedError

In [12]:
from typing import Tuple

def displayed_image(
        i_cell: np.ndarray,
        vdd: float,
        p1: float = 4.251e-5,
        p2: float = -3.029e-4,
        p3: float = 3.024e-5,
        orig_vdd: float = 15,
        ) -> Tuple[np.ndarray, np.ndarray]:
    """
    Display an image on the OLED display taking into account the effect of DVS.

    :param i_cell: An array of the currents drawn by each pixel of the display.
    :param vdd: The new voltage of the display.
    """
    i_cell_max = (p1 * vdd * 1) + (p2 * 1) + p3
    image_rgb_max = (i_cell_max - p3) / (p1 * orig_vdd + p2) * 255
    out = np.round((i_cell - p3) / (p1 * orig_vdd + p2) * 255)
    original_image = out.copy()

    # Clip the values exceeding `i_cell_max` to `image_rgb_max`
    out[i_cell > i_cell_max] = image_rgb_max

    return original_image.astype(np.uint8), out.astype(np.uint8)

In [ ]:
from scipy.io import loadmat

# Load the .mat file
mat_data = loadmat('sample_cell_current.mat')["I_cell_sample"]

# Inspect the loaded data
print(f"Shape: {mat_data.shape}")
print(f"Red channel:\n{mat_data[:, :, 0]}")
print(f"Green channel:\n{mat_data[:, :, 1]}")
print(f"Blue channel:\n{mat_data[:, :, 2]}")

image_array_orig, image_array_w_dvs = displayed_image(mat_data, 10)
image_orig = Image.fromarray(image_array_orig)
image_orig.show()
image_w_dvs = Image.fromarray(image_array_w_dvs)
image_w_dvs.show()


Shape: (176, 220, 3)
Red channel:
[[2.46842941e-04 2.24526275e-04 2.03522353e-04 ... 9.45645098e-05
  9.45645098e-05 9.45645098e-05]
 [2.46842941e-04 2.24526275e-04 2.03522353e-04 ... 9.45645098e-05
  9.45645098e-05 9.45645098e-05]
 [2.46842941e-04 2.24526275e-04 2.03522353e-04 ... 9.45645098e-05
  9.45645098e-05 9.45645098e-05]
 ...
 [2.88850784e-04 2.78348824e-04 2.67846863e-04 ... 3.32171373e-04
  3.21669412e-04 3.21669412e-04]
 [2.88850784e-04 2.78348824e-04 2.67846863e-04 ... 3.32171373e-04
  3.32171373e-04 3.21669412e-04]
 [2.88850784e-04 2.78348824e-04 2.67846863e-04 ... 3.32171373e-04
  3.32171373e-04 3.11167451e-04]]
Green channel:
[[1.41823333e-04 1.20819412e-04 1.10317451e-04 ... 7.22478431e-05
  6.69968627e-05 6.69968627e-05]
 [1.36572353e-04 1.26070392e-04 1.10317451e-04 ... 7.22478431e-05
  6.69968627e-05 6.69968627e-05]
 [1.31321373e-04 1.20819412e-04 1.10317451e-04 ... 7.22478431e-05
  6.69968627e-05 6.69968627e-05]
 ...
 [1.73329216e-04 1.62827255e-04 1.57576275e-04 ..